feat: generic Grid iterator

CarloVarni opened the pull request:
Just an idea @andiwand @paulgessinger @LuisFelipeCoelho @noemina 

This adds two grid iterators:
- A global iterator that simply runs on the global index. As such you also run on the under/over flow bins
- A local iterator that uses instead local indexes. This one avoids under/over flow bins

The cool thing about the local one is that is also allows for a custom navigation pattern for any axis. This is currently done for the Z azis where we jump from the first bins, then the last bins and then the central Z bins in the grid

Still a draft since I need to add a few things. But opening anyway to get additional feedbacks

LuisFelipeCoelho reviewed (changes requested):
Do you have any idea of the effect on the time performance?

CarloVarni commented:
> Do you have any idea of the effect on the time performance?

basically irrelevant. Iterating on the 2D grid we use for seeding increase from 6 to 12 micro-seconds. Compared to the rest of the seeding this is nothing imho

paulgessinger commented:
@niermann999 would you be able to review this?

CarloVarni commented:
There are still a couple of functionalities I need to add following the recent developments for fast tracking. Can the review be done after I implement those?

CarloVarni commented:
ok @niermann999 I think we are ready for review

CarloVarni commented:
@LuisFelipeCoelho note that with this change the `skipZMiddleBin` property will almost become pointless. The iterator will be able to iterate on whatever sequence of bins you like.

In other words... for a usual seed finding process we have:
```python
zBinsCustomLooping = [1, 2, 3, 4, 11, 10, 9, 8, 6, 5, 7]
```
but for fast tracking you can simply provide a shorter vector. For instance:
```python
zBinsCustomLooping = [11, 10, 9, 8, 6, 5, 7]
```
much simpler imho.

As of now I have not removed yet the `skipZMiddleBin` property since I do not want to make this PR a breaking one. But I'm planning to do it in a future PR

niermann999 commented on Core/include/Acts/Utilities/GridIterator.ipp:
```diff
@@ -0,0 +1,282 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include <iostream>
```
leftover?

CarloVarni replied:
ops ...

niermann999 commented on Tests/UnitTests/Core/Utilities/GridIterationTests.cpp:
```diff
@@ -0,0 +1,373 @@
+  BOOST_CHECK_EQUAL(std::distance(gridStart, gridStop), 3ul);
+  BOOST_CHECK_EQUAL(gridStart != gridStop, true);
+
+  [[maybe_unused]] double value = *gridStart;
```
Isn't this, in  fact, always unused?

CarloVarni replied:
yes indeed. This is for testing the `operator*` doesn't have sudden crushes. We do nothing with `value`

As far as I know c++ has this `maybe_unused` attribute. Is there as well a `unused` attribute, too? If so, I'm not aware of it

niermann999 commented on Tests/UnitTests/Core/Utilities/GridIterationTests.cpp:
```diff
@@ -0,0 +1,373 @@
+
+  BOOST_CHECK_EQUAL(numIterations, expectedIterations);
+}
+
```
Is there also a test that accesses values and checks that the iterator hits all of the right positions? Or is it enough that the seeding tests pass?

CarloVarni replied:
I can add something of this kind indeed

CarloVarni replied:
I have added a test for it

niermann999 commented:
Since I ran into that myself, does this work correctly on grids with phi-axes, where the local indices wrap around or is that not an issue?

CarloVarni commented on Core/include/Acts/Seeding/BinnedSPGroup.hpp:
```diff
@@ -39,9 +40,12 @@ class BinnedSPGroupIterator {
   BinnedSPGroupIterator(BinnedSPGroup<external_spacepoint_t>&& group,
-                        std::size_t) = delete;
+                        std::array<std::size_t, 2> index,
+                        std::array<std::vector<std::size_t>, 2> navigation) =
```
totally unrelated.

In this context `navigation` defines a sequence of bins in the axis so that you can jump from one to another when iterating in the grid. This is what we currently do in the seeding for the Z-Axis

CarloVarni commented on Core/include/Acts/Seeding/BinnedSPGroup.ipp:
```diff
@@ -12,40 +12,32 @@
+  endline[0ul] = navigation[0ul].size();
+  endline[1ul] = navigation[1ul].size();
+  m_gridItrEnd =
+      typename Acts::SpacePointGrid<external_spacepoint_t>::local_iterator_t(
```
yes, but I preferd to be explicit in this case

CarloVarni commented:
Hi @niermann999 I should have addressed all your comments. Can you take another look?

Concerning your question on phi-axes, that should not be a problem in this case. That will be handled when looking for neighboring bins, which is done in another class

niermann999 reviewed (approved):
Did you check that this works correctly at the edge of an axis, in particular a phi axis or is it just up to the user to pass the correct navigation?

acts-project-service commented:
## :white_check_mark: Athena integration test results
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765079)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsBenchmarkWithSpot.sh 8 100](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765078)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765077)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateAmbiguityResolution.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765076)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateResolvedTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765075)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765074)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765073)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765072)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765071)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765070)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765069)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765068)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsGSFRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765067)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765066)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765065)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765064)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765063)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765062)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765061)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765060)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765059)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765058)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34765057)||
